tests/sagemaker/test_sagemaker_deployment_client.py

```diff
@@ -0,0 +1,722 @@
+# @mock_sagemaker_aws_services
+# def test_deploy_replace_in_asynchronous_mode_returns_before_endpoint_creation_completes(
+#     pretrained_model, sagemaker_client, sagemaker_deployment_client
+# ):
+#     endpoint_update_latency = 10
+#     get_sagemaker_backend(sagemaker_client.meta.region_name).set_endpoint_update_latency(
+#         endpoint_update_latency
+#     )
+
+#     name = "test-app"
+#     sagemaker_deployment_client.create_deployment(
+#         name=name,
+#         model_uri=pretrained_model.model_uri,
+#         config=dict(mode=mfs.DEPLOYMENT_MODE_CREATE, synchronous=True),
+#     )
+
+#     update_start_time = time.time()
+#     sagemaker_deployment_client.create_deployment(
+#         name=name,
+#         model_uri=pretrained_model.model_uri,
+#         config=dict(mode=mfs.DEPLOYMENT_MODE_REPLACE, synchronous=False, archive=True),
+#     )
+#     update_end_time = time.time()
+
+#     assert (update_end_time - update_start_time) < endpoint_update_latency
+#     endpoint_description = sagemaker_client.describe_endpoint(EndpointName=name)
+#     assert endpoint_description["EndpointStatus"] == Endpoint.STATUS_UPDATING
+
+
+@pytest.mark.large
+@mock_sagemaker_aws_services
+def test_create_deployment_throws_exception_after_endpoint_creation_fails(
+    pretrained_model, sagemaker_client, sagemaker_deployment_client
+):
+    endpoint_creation_latency = 10
+    sagemaker_backend = get_sagemaker_backend(sagemaker_client.meta.region_name)
+    sagemaker_backend.set_endpoint_update_latency(endpoint_creation_latency)
+
+    boto_caller = botocore.client.BaseClient._make_api_call
+
+    def fail_endpoint_creations(self, operation_name, operation_kwargs):
+        """
+        Processes all boto3 client operations according to the following rules:
+        - If the operation is an endpoint creation, create the endpoint and set its status to
+          ``Endpoint.STATUS_FAILED``.
+        - Else, execute the client operation as normal
+        """
+        result = boto_caller(self, operation_name, operation_kwargs)
+        if operation_name == "CreateEndpoint":
+            endpoint_name = operation_kwargs["EndpointName"]
+            sagemaker_backend.set_endpoint_latest_operation(
+                endpoint_name=endpoint_name,
+                operation=EndpointOperation.create_unsuccessful(
+                    latency_seconds=endpoint_creation_latency
+                ),
+            )
+        return result
+
+    with mock.patch(
+        "botocore.client.BaseClient._make_api_call", new=fail_endpoint_creations
+    ), pytest.raises(MlflowException, match="deployment operation failed") as exc:
+        sagemaker_deployment_client.create_deployment(
+            name="test-app",
+            model_uri=pretrained_model.model_uri,
+        )
+
+    assert "deployment operation failed" in exc.value.message
+    assert exc.value.error_code == ErrorCode.Name(INTERNAL_ERROR)
+
+
+# @pytest.mark.large
+# @mock_sagemaker_aws_services
```

Comment:
Let me know if I can help debug the commented tests or if they're a work in progress or just used for debugging during your implementation.